optuna/_optimize.py

```diff
@@ -193,6 +196,20 @@ def _run_trial(
     func_err_fail_exc_info: Optional[Any] = None
     # Set to a string if `func` returns correctly but the return value violates assumptions.
     values_conversion_failure_message: Optional[str] = None
+    stop_event: Optional[Event] = None
+    thread: Optional[Thread] = None
+
+    if (
+        study._storage.is_heartbeat_supported()
+        and study._storage.get_heartbeat_interval() is not None
+    ):
+        study._storage.record_heartbeat(trial._trial_id)
+        stop_event = Event()
+        thread = Thread(
+            target=_record_heartbeat, args=(trial._trial_id, study._storage, stop_event)
+        )
+        thread.start()
+        study._storage.fail_stale_trials()
```

Comment:
How about doing this before starting the heartbeat thread for this trial? Or do we really want to ever fail the current trial here?